zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -984,11 +997,50 @@ private void fireAllEvents() {
 
   protected List<String> getChildren(final String path, final boolean watch) {
     long startT = System.currentTimeMillis();
+
     try {
       List<String> children = retryUntilConnected(new Callable<List<String>>() {
+        private int connectionLossRetryCount = 0;
+
         @Override
         public List<String> call() throws Exception {
-          return getConnection().getChildren(path, watch);
+          try {
+            return getConnection().getChildren(path, watch);
+          } catch (ConnectionLossException e) {
+            ++connectionLossRetryCount;
+            // Allow retrying 3 times before checking stat checking number of children,
+            // because there is a higher possibility that connection loss is caused by other
+            // factors such as network connectivity, connected ZK node could not serve
+            // the request, session expired, etc.
+            if (connectionLossRetryCount >= 3) {
+              // Issue: https://github.com/apache/helix/issues/962
+              // Connection loss might be caused by an excessive number of children.
+              // Infinitely retrying connecting may cause high GC in ZK server and kill ZK server.
+              // This is a workaround to check numChildren to have a chance to exit retry loop.
+              // TODO: remove this check once we have a better way to exit infinite retry
+              Stat stat = getStat(path);
+              if (stat != null) {
+                if (stat.getNumChildren() > NUM_CHILDREN_LIMIT) {
+                  LOG.error("Failed to get children for path {} because number of children {} "
+                          + "exceeds limit {}, aborting retry.", path, stat.getNumChildren(),
+                      NUM_CHILDREN_LIMIT);
+                  // MarshallingErrorException could represent transport error: exceeding the
+                  // Jute buffer size. So use it to exit retry loop and tell that zk is not able to
+                  // transport the data because packet length is too large.
+                  throw new KeeperException.MarshallingErrorException();
+                } else {
+                  // No need to do stat again for next connection loss.
+                  connectionLossRetryCount = 0;
```

Comment:
I think you might want to keep checking the children's number on any retries after the first 3. Or, what if the children are increased during the retry, right?

So just make the logic like,
if (count < 3 || getChildrenCount() < limit) {
  count++;
  retry;
} else {
  exist;
}

Reply:
> Your logic here is only allow retrying 3 times, after 3 retries, exit retry loop.

Could you please check my logic carefully. It will keep retrying if the ChildrenCount is good. It is "||", not "&&". It is doing almost the same as your logic. getChildrenCount() won't be called if the count is smaller than 3.
This is helping you to simplify the code.

Reply:
> I think you might want to keep checking the children's number on any retries after the first 3. Or, what if the children are increased during the retry, right?

Yes, it will still check numChildren after the first 3 retries. The strategy is check numChildren every other 3 retries.
Your code checks numChildren every time after the first 3 retries, right? But actually I don't think it is necessary to check every time after 3 retries. I would like to just check numChildren every other 3 retries. So maybe your code could not really simplify the logic. I'll sync up with you.